include/rspq.h

```diff
@@ -0,0 +1,716 @@
+ * 
+ * The net effect is that commands enqueued in the high-priority queue are
+ * executed right away by the RSP, irrespective to whatever was currently
+ * enqueued in the standard queue. This can be useful for running tasks that
+ * require immediate execution, like for instance audio processing.
+ * 
+ * If required, it is possible to call #rspq_highpri_sync to wait for the
+ * high-priority queue to be fully executed.
+ * 
+ * Notice that the RSP cannot be fully preempted, so switching to the high-priority
+ * queue can only happen after a command has finished execution (before starting
+ * the following one). This can have an effect on latency if a single command
+ * has a very long execution time; RSP overlays should in general prefer
+ * providing smaller, faster commands.
+ * 
+ * This feature should normally not be used by end-users, but by libraries
+ * in which a very low latency of RSP execution is paramount to their workings.
+ * 
+ */
+
+#ifndef __LIBDRAGON_RSPQ_H
+#define __LIBDRAGON_RSPQ_H
+
+#include <stdint.h>
+#include <rsp.h>
+#include <pputils.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** @brief Maximum size of a command (in 32-bit words). */
+#define RSPQ_MAX_COMMAND_SIZE          16
+
+
+/**
+ * @brief A preconstructed block of commands
+ * 
+ * To improve performance of execution of sequences of commands, it is possible
+ * to create a "block". A block is a fixed set of commands that is created
+ * once and executed multiple times.
+ * 
+ * To create a block, use #rspq_block_begin and #rspq_block_end. After creation,
+ * you can use #rspq_block_run at any point to run it. If you do not need the
+ * block anymore, use #rspq_block_free to dispose it.
+ */
+typedef struct rspq_block_s rspq_block_t;
+
+/**
+ * @brief A syncpoint in the queue
+ * 
+ * A syncpoint can be thought of as a pointer to a position in the command queue.
+ * After creation, it is possible to later check whether the RSP has reached
+ * that position or not.
+ * 
+ * To create a syncpoint, use #rspq_syncpoint that returns a syncpoint that
+ * references the current position. Call #rspq_check_syncpoint or #rspq_wait_syncpoint
+ * to respectively do a single check or block waiting for the syncpoint to be
+ * reached by RSP.
+ * 
+ * Syncpoints are implemented using interrupts, so they have a light but non
+ * trivial overhead. Do not abuse them. For instance, it is reasonable to use
+ * tens of syncpoints per frame, but not hundreds or thousands of them.
+ * 
+ * @note A valid syncpoint is an integer greater than 0.
+ */
+typedef int rspq_syncpoint_t;
+
+/**
+ * @brief Initialize the RSPQ library.
+ * 
+ * This should be called by the initialization functions of the higher-level
+ * libraries using the RSP command queue. It can be safely called multiple
+ * times without side effects.
+ * 
+ * It is not required by applications to call this explicitly in the main
+ * function.
+ */
+void rspq_init(void);
```

Comment:
One important thing to note; once the system is initialized the cpu should not DMA anything to the RSP unless it is known that the rspq is halted on the CPU, which can be achieved by `rspq_wait`.
This can include a scenario where the user wants to switch over to a custom ucode for some reason. There is no DMA synchronization and doing a DMA while rspc is running is UB.
Might be obvious but wanted to point it out anyways :)

Reply:
Yeah pretty much. If the user wants to load some custom ucode when rspq is running, they have to do the synchronization themselves.

This made me think - should we check whether rspq is actually running when the user accesses the API? Or maybe just add a function `rspq_is_running` or something so other components that depend on it (like the mixer) can perform this check? Or we document the dependency on those components to warn the user about calling `rspq_close` before the dependent components have been closed properly.

Reply:
So ideally, you would just call `rspq_stop`, load your own ucode and run it, and then call `rspq_start` and it would just be ready to go from where it left off, without the need to re-register overlays again and such. I think a reasonable restriction is that there cannot be any commands in the queue while stopping that will resume executing after restarting. So internally it would first call `rspq_wait` and then maybe load some dummy overlay so the current overlay's state is saved. I can look into implementing this.

Reply:
I think we don't need the pause functionality at this time and we can develop it later. For now, it should be ok to just have some warnings to note manual handling is necessary for opting out of rspq. Using mixer also means it will be inited so a small note there would be nice as well.

Reply:
It's simple in theory, but in practice I'd have to do some more plumbing to make it work which is not entirely trivial. For example we would need to consider the case in which all overlay slots are occupied. Then one of those overlays needs to be temporarily unregistered or somehow backed up, then the dummy overlay takes its place to trigger the state-saving mechanism. Also things like the pointer stack and the current RDRAM address need to be backed up from DMEM.

I've decided to leave this for a follow up PR for now.